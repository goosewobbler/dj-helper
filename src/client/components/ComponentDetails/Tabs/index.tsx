import * as React from 'react';

interface TabsProps {
  children?: any;
  headings: any[];
  renderButtons: any;
}

interface TabsState {
  selectedIndex: number;
}

class Tabs extends React.Component<TabsProps, TabsState> {
  constructor(props: TabsProps) {
    super(props);

    this.state = {
      selectedIndex: 0,
    };
  }

  public render() {
    const panels = Array.isArray(this.props.children) ? this.props.children : [this.props.children];

    return (
      <div>
        <div className="header">
          <ul>{this.renderHeadings()}</ul>
          {this.props.renderButtons()}
        </div>
        <div>{panels[this.state.selectedIndex]}</div>
      </div>
    );
  }

  private renderHeadings() {
    return this.props.headings.map((heading, index) => {
      const selected = this.state.selectedIndex === index;
      const onClick = () => {
        this.setState({
          selectedIndex: index,
        });
      };
      return (
        <button key={heading} onClick={onClick}>
          {heading}
        </button>
      );
    });
  }
}

export default Tabs;
