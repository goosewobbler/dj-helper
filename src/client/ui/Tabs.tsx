import * as React from 'react';

interface ITabsProps {
  children?: any;
  headings: any[];
  buttons?: {
    render: any;
    props: any;
  };
}

interface ITabsState {
  selectedIndex: number;
}

class Tabs extends React.Component<ITabsProps, ITabsState> {
  constructor(props: ITabsProps) {
    super(props);

    this.state = {
      selectedIndex: 0,
    };
  }

  public render() {
    const panels = Array.isArray(this.props.children) ? this.props.children : [this.props.children];

    return (
      <div>
        <div className="header">
          <ul>{this.renderHeadings()}</ul>
          {this.props.buttons.render(this.props.buttons.props)}
        </div>
        <div>{panels[this.state.selectedIndex]}</div>
      </div>
    );
  }

  private renderHeadings() {
    return this.props.headings.map((heading, index) => {
      const selected = this.state.selectedIndex === index;
      const onClick = () => {
        this.setState({
          selectedIndex: index,
        });
      };
      return (
        <button key={heading} onClick={onClick}>
          {heading}
        </button>
      );
    });
  }
}

export default Tabs;
