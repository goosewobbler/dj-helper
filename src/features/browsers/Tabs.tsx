import React, { ReactElement } from 'react';
import { Tab } from '@headlessui/react';
import { useAppDispatch, useAppSelector } from '../../common/hooks';
import { createBrowser, selectActiveBrowser, tabClosed, tabSelected } from './browsersSlice';
import { Browser } from '../../common/types';
import { MetaPanel } from './MetaPanel';

function classNames(...classes: string[]) {
  return classes.filter(Boolean).join(' ');
}

export const Tabs = ({ browsers }: { browsers: Browser[] }): ReactElement => {
  const dispatch = useAppDispatch();
  const activeBrowser = useAppSelector(selectActiveBrowser());
  const isSelected = (id: number) => id === activeBrowser.id;

  return (
    <Tab.Group
      defaultIndex={activeBrowser.id}
      onChange={(index) => {
        dispatch(tabSelected({ id: index }));
      }}
    >
      <div className="flex justify-between" data-testid="header">
        <Tab.List>
          {browsers.map(
            (browser: Browser, index: number): ReactElement => (
              <Tab key={browser.id}>
                {() => (
                  <button
                    type="button"
                    className={classNames(
                      isSelected(browser.id) ? 'text-gray-900' : 'text-gray-500 hover:text-gray-700',
                      'rounded-r-lg rounded-l-lg',
                      'group relative min-w-0 flex-1 overflow-hidden bg-white py-4 px-4 text-sm font-medium text-center hover:bg-gray-50 focus:z-10',
                    )}
                  >
                    <span>
                      {browser.title}
                      <button type="button" onClick={() => dispatch(tabClosed({ id: index }))}>
                        x
                      </button>
                    </span>
                    <span
                      aria-hidden="true"
                      className={classNames(
                        isSelected(browser.id) ? 'bg-indigo-500' : 'bg-transparent',
                        'absolute inset-x-0 bottom-0 h-0.5',
                      )}
                    />
                  </button>
                )}
              </Tab>
            ),
          )}
          <button
            className="p-1 mx-2 my-0 text-lg border-0 border-solid outline-none text-primary-text"
            type="button"
            onClick={() => dispatch(createBrowser({}))}
          >
            +
          </button>
        </Tab.List>
      </div>
      <div className="relative flex flex-col flex-grow w-full h-full" data-testid="panels">
        <Tab.Panels>
          {browsers.map(
            (browser: Browser): ReactElement => (
              <Tab.Panel key={browser.id}>
                <MetaPanel browser={browser} />
              </Tab.Panel>
            ),
          )}
        </Tab.Panels>
      </div>
    </Tab.Group>
  );
};
