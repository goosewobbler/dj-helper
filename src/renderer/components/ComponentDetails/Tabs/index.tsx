import * as React from 'react';

interface TabsProps {
  children?: React.ReactElement[];
  headings: string[];
  renderButtons(): React.ReactElement;
}

interface TabsState {
  selectedIndex: number;
}

class Tabs extends React.Component<TabsProps, TabsState> {
  public constructor(props: TabsProps) {
    super(props);

    this.state = {
      selectedIndex: 0,
    };
  }

  private renderHeadings(): React.ReactElement[] {
    const { selectedIndex } = this.state;
    const { headings } = this.props;
    return headings.map(
      (heading, index): React.ReactElement => {
        const selected = selectedIndex === index;
        const borderBottomWidth = `border-b-${selected ? '4' : '2'}`;
        const borderBottomColor = `border-${selected ? 'selected-item' : 'primary-text'}`;
        const onClick = (): void => {
          this.setState({
            selectedIndex: index,
          });
        };
        return (
          <button
            className={`p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid ${borderBottomWidth} ${borderBottomColor}`}
            type="button"
            key={heading}
            onClick={onClick}
          >
            {heading}
          </button>
        );
      },
    );
  }

  public render(): React.ReactElement {
    const { children, renderButtons } = this.props;
    const { selectedIndex } = this.state;
    const panels = Array.isArray(children) ? children : [children];

    return (
      <div className="flex flex-col flex-grow">
        <div className="flex flex-row justify-between header">
          <ul className="flex px-0 pt-2 pb-4">{this.renderHeadings()}</ul>
          {renderButtons()}
        </div>
        <div className="relative flex flex-col flex-grow w-full h-full">{panels[selectedIndex]}</div>
      </div>
    );
  }
}

export default Tabs;
