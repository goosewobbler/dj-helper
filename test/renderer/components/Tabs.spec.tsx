import * as React from 'react';
import { render, RenderResult, fireEvent } from '@testing-library/react';
import Tabs from '../../../src/renderer/components/ComponentDetails/Tabs';

let tabs: RenderResult;

afterEach(() => {
  tabs.unmount();
});

describe('given a list of tab headings', () => {
  let headings: string[];

  beforeEach(() => {
    headings = ['look', 'some', 'headings'];
    tabs = render(
      <Tabs headings={headings}>
        <div className="child">I am child 1</div>
        <div className="child">I am child 2</div>
        <div className="child">I am child 3</div>
      </Tabs>,
    );
  });

  it('should render tab buttons with the expected text', () => {
    expect(tabs.getByTestId('header')).toContainElement(tabs.getByText('look'));
    expect(tabs.getByTestId('header')).toContainElement(tabs.getByText('some'));
    expect(tabs.getByTestId('header')).toContainElement(tabs.getByText('headings'));
  });

  it('should render the expected panel', () => {
    expect(tabs.getByTestId('panels')).toContainElement(tabs.getByText('I am child 1'));
  });

  it('should render the expected html', () => {
    expect(tabs.container).toMatchSnapshot();
  });

  describe('when a tab button is clicked', () => {
    beforeEach(() => {
      const button = tabs.getByText('headings');
      fireEvent.click(button);
    });

    it('should render the expected panel', () => {
      expect(tabs.getByTestId('panels')).toContainElement(tabs.getByText('I am child 3'));
    });

    it('should render the expected html', () => {
      expect(tabs.container).toMatchSnapshot();
    });
  });
});

describe('and heading children are provided', () => {
  beforeEach(() => {
    tabs = render(
      <Tabs headings={['so', 'many', 'headings']} headingChildren={<div>I am heading child</div>}>
        <div className="child">I am child 1</div>
        <div className="child">I am child 2</div>
        <div className="child">I am child 3</div>
      </Tabs>,
    );
  });

  it('should render the expected heading children', () => {
    expect(tabs.getByTestId('header')).toContainElement(tabs.getByText('I am heading child'));
  });

  it('should render the expected html', () => {
    expect(tabs.container).toMatchSnapshot();
  });
});
