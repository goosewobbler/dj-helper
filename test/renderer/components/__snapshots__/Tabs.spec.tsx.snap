// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`and heading children are provided should render the expected html 1`] = `
<div>
  <div
    class="flex justify-between"
    data-testid="header"
  >
    <ul
      class="flex px-0 pt-2 pb-4"
    >
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-4 border-selected-item"
        type="button"
      >
        so
      </button>
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-2 border-primary-text"
        type="button"
      >
        many
      </button>
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-2 border-primary-text"
        type="button"
      >
        headings
      </button>
    </ul>
    <div>
      I am heading child
    </div>
  </div>
  <div
    class="relative flex flex-col flex-grow w-full h-full"
    data-testid="panels"
  >
    <div
      class="child"
    >
      I am child 1
    </div>
  </div>
</div>
`;

exports[`given a list of tab headings should render the expected html 1`] = `
<div>
  <div
    class="flex justify-between"
    data-testid="header"
  >
    <ul
      class="flex px-0 pt-2 pb-4"
    >
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-4 border-selected-item"
        type="button"
      >
        look
      </button>
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-2 border-primary-text"
        type="button"
      >
        some
      </button>
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-2 border-primary-text"
        type="button"
      >
        headings
      </button>
    </ul>
  </div>
  <div
    class="relative flex flex-col flex-grow w-full h-full"
    data-testid="panels"
  >
    <div
      class="child"
    >
      I am child 1
    </div>
  </div>
</div>
`;

exports[`given a list of tab headings when a tab button is clicked should render the expected html 1`] = `
<div>
  <div
    class="flex justify-between"
    data-testid="header"
  >
    <ul
      class="flex px-0 pt-2 pb-4"
    >
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-2 border-primary-text"
        type="button"
      >
        look
      </button>
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-2 border-primary-text"
        type="button"
      >
        some
      </button>
      <button
        class="p-1 outline-none border-0 text-lg my-0 mx-2 text-primary-text border-solid border-b-4 border-selected-item"
        type="button"
      >
        headings
      </button>
    </ul>
  </div>
  <div
    class="relative flex flex-col flex-grow w-full h-full"
    data-testid="panels"
  >
    <div
      class="child"
    >
      I am child 3
    </div>
  </div>
</div>
`;
